book/src/rfcs/foreign_field_mul.md

```diff
@@ -0,0 +1,1273 @@
+\end{aligned}
+$$
+
+Remember $f' = 2^t - f$ is our negated foreign field modulus.  Thus, we have
+
+$$
+\begin{aligned}
+f' \le q &+ f' < 2^t \\
+\end{aligned}
+$$
+
+So to check $q < t$ we just need to compute $q' = q + f'$ and check $f' \le q' < 2^t$
+
+Observe that
+
+$$
+0 \le q \implies  f' \le q'
+$$
+
+and that
+
+$$
+q' < 2^t \implies q < f
+$$
+
+So we only need to check
+
+- $0 \le q$
+- $q' < 2^t$
+
+The first check is always true since we are operating over the positive integers and $q \in \mathbb{Z^+}$.  Observe that the second check also constrains that $q < 2^t$, since $f \le 2^{259} < 2^t$ and thus
+
+$$
+\begin{aligned}
+q' &\le 2^t \\
+q + f' &\le 2^t \\
+q &\le 2^t - (2^t - f) = f\\
+q &< 2^t
+\end{aligned}
+$$
+
+Therefore, to constrain $q < f$ we only need constraints for
+
+- $q' = q + f'$
+- $q' < 2^t$
+
+and we don't require an additional multi-range-check for $q < 2^t$.
+
+### Cost of bound checks
+
+This section analyzes the structure and costs of bounds checks for foreign field addition and foreign field multiplication.
+
+#### Addition
+
+In our foreign field addition design the operands $a$ and $b$ do not need to be less than $f$.  The field overflow bit $\mathcal{o}$ for foreign field addition is at most 1.  That is, $a + b = \mathcal{o} \cdot f + r$, where $r$ is allowed to be greater than $f$.  Therefore,
+
+$$
+(f + a) + (f + b) = 1 \cdot f + (f + a + b)
+$$
+
+These can be chained along $k$ times as desired.  The final result
+
+$$
+r = (\underbrace{f + \cdots + f}_{k} + a_1 + b_1 + \cdots a_k + b_k)
+$$
+
+Since the bit length of $r$ increases logarithmically with the number of additions, in Kimchi we must only check that the final $r$ in the chain is less than $f$ to constrain the entire chain.
+
+> **Security note:** In order to defer the $r < f$ check to the end of any chain of additions, it is extremely important to consider the potential impact of wraparound in $\mathbb{F_n}$.  That is, we need to consider whether the addition of a large chain of elements greater than the foreign field modulus could wrap around.  If this could happen then the $r < f$ check could fail to detect an valid witness.  Below we will show that this is not possible in Kimchi.
+>
+> Recall that our foreign field elements are comprised of 3 limbs of 88-bits each that are each represented as native field elements in our proof system.  In order to wrap around and circumvent the $r < f$ check, the highest limb would need to wrap around.  This means that an attacker would need to perform about $k \approx n/2^{\ell}$ additions of elements greater than then foreign field modulus.  Since Kimchi's native moduli (Pallas and Vesta) are 255-bits, the attacker would need to provide a witness for about $k \approx 2^{167}$ additions.  This length of witness is greater than Kimchi's maximum circuit (resp. witness) length.  Thus, it is not possible for the attacker to generate a false proof by causing wraparound with a large chain of additions.
```

Comment:
Yeah! Then the idea about not fitting in the circuit worked out!

Reply:
Yey!